### ISSUE TITLE
R2/L2 button maps not working on ViGEm DualShock4

### ISSUE THREAD

**Issue by TimAkgayev** [Snoothy/UCR#76]
Mapping to R2 or L2 does not trigger them in game (PSNow).

**Comment by evilC** [comment]
L2 and R2 are not logically "Buttons", they are axes.  
When you connect a real DS4 to a PC and press the L2 trigger, then it triggers an axis *and* a button.  
However, this is just for the DirectInput representation of a DS4 pad.  
I am unsure if, while plugged into a PS4 console, there is actually any such thing as a button for L2.  
Many games may let you treat it as a button, but that does not mean it actually is one.  
TLDR - You should try sending output to the **Axis** L2 instead of the **Button** L2.  
If you need to map a button (eg keyboard key) to L2, then use a ButtonToAxis plugin

**Comment by habeelmazhar** [comment]
![screenshot 1](https://user-images.githubusercontent.com/7492074/51409095-a8035580-1b82-11e9-9a99-1fa89f402053.png)

is this an issue that LT and  RT does not work properly? Because in the preview and html5 tester it does not respond properly e.g the value in html tester is 1.00 on press and on release its gets stuck to 1.00 as you can see the values are properly 0 and 100.

**Comment by evilC** [comment]
It is an issue to do with the emulation.  
If you plug a real DS4 into a PC, when you press a trigger, BOTH a button AND and axis change state.  
I can't remember the exact details, but for argument's sake, lets say that When you press L2, then both button 5 and axis 5 change state.  

It is impossible, with a real DS4, to move axis 5 without making button 5 light up.  
However, in ViGEm, this IS possible - setting button 5 to pressed does NOT set the state of axis 5  

As far as I am aware, most DS4 stuff on PC will only pay attention to the state of axis 5 and ignore button 5, so in general the solution is simply to use a ButtonToAxis instead of a ButtonToButton.  
If you need it to totally replicate how a real DS4 works, then add both a ButtonToButton and a ButtonToAxis

**Comment by TimAkgayev** [comment]
@evilC I'm using a PS2 controller with a USB adapter to map to a DualShock 4. I've tried to switch the L2 and R2 to use Button To Axis plugin (Axis to Axis doesn't work because I think on PS2 L2/R2 were still buttons). The thing is when I switch it there is weird behaviour. For example the R2 works once. If I press it right after I load the profile the character begins to "charge" the attack however when I let go of R2 the game doesn't register the release and just keeps the character in "charging to attack" mode. If I press other buttons it gets out of that state, however the next time I press R2 nothing happens at all. I'm guessing it somehow doesn't return to released state. Now, if I go into a menu (in this game menu tabs are switched by L2/R2) and try to switch tabs with then I can switch once with L2 and once with R2 but after I press each button once it stops registering following presses.

**Comment by evilC** [comment]
Button**s**ToAxis is for mapping two buttons to one *centered* axis. You want the ButtonToAxis plugin with a press value of 100 and a release value of -100

**Comment by TimAkgayev** [comment]
Setting it to -100 fixed the problem, thanks.
